### PR TITLE
symcc: init at 1.0-unstable-2024-07-16

### PR DESCRIPTION
**Pull request by jollheef**
Adds symcc, compiler-based symbolic execution engine.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sy/symcc/package.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  postPatch = ''
+    echo 'install(TARGETS SymCC LIBRARY DESTINATION '\$'{CMAKE_INSTALL_LIBDIR})' >> CMakeLists.txt
+  '';
+
+  postInstall = ''
+    cp SymCCRuntime-prefix/src/SymCCRuntime-build/libsymcc-rt.{a,so} $out/lib/
+
+    mkdir -p $out/bin/
+    cp symcc $out/bin/
+    cp sym++ $out/bin/
```
Comment (sikmir) [inline pkgs/by-name/sy/symcc/package.nix:39]:
```suggestion
    install -Dm755 symcc sym++ -t $out/bin
```

File: pkgs/by-name/sy/symcc/package.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "symcc";
+  version = "1.0-unstable-2024-07-16";
+
+  src = fetchFromGitHub {
+    owner = "eurecom-s3";
+    repo = "symcc";
+    rev = "65a3633992318ded8939629eda54022932fd582d";
+    sha256 = "sha256-WzZLeq+if7FyQKkSMDiqxnjtbb7eg8EVuNkCwxyEryM=";
```
Comment (sikmir) [inline pkgs/by-name/sy/symcc/package.nix:21]:
```suggestion
    hash = "sha256-WzZLeq+if7FyQKkSMDiqxnjtbb7eg8EVuNkCwxyEryM=";
```

File: pkgs/by-name/sy/symcc/package.nix [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  fetchFromGitHub,
+  cmake,
+  ninja,
+  llvm_17,
+  clang_17,
+  z3,
+  makeWrapper,
+}:
+
+stdenv.mkDerivation rec {
```
Comment (sikmir) [inline pkgs/by-name/sy/symcc/package.nix:13]:
```suggestion
stdenv.mkDerivation {
```

**Comment by jollheef** [comment]
Thank you. Fixed.

File: pkgs/by-name/sy/symcc/package.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  nativeBuildInputs = [
+    cmake
+    ninja
+    llvm_17
+    clang_17
+    makeWrapper
+  ];
+
+  buildInputs = [
+    z3
```
Comment (FliegendeWurst) [inline pkgs/by-name/sy/symcc/package.nix:57]:
```suggestion
    llvm_17
    z3
```
Otherwise the build fails if you set strictDeps = true.

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sy/symcc/package.nix (hunk 4 above)
Comment (jollheef) [reply pkgs/by-name/sy/symcc/package.nix:58]: Thanks.

**Review by FliegendeWurst — APPROVED** [review]

File: pkgs/by-name/sy/symcc/package.nix [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    clang_17
+    makeWrapper
+  ];
+
+  buildInputs = [
+    llvm_17
+    z3
+  ];
+
+  meta = {
```
Comment (FliegendeWurst) [inline pkgs/by-name/sy/symcc/package.nix:61]: You may want to set platforms = linux, or fix the build error
```
/tmp/nix-build-symcc-1.0-unstable-2024-07-16.drv-0/source/runtime/src/backends/qsym/qsym/qsym/pintool/solver.cpp:2:10: fatal error: 'byteswap.h' file not found
    2 | #include <byteswap.h>
      |          ^~~~~~~~~~~~
1 error generated.
```

Comment (jollheef) [reply pkgs/by-name/sy/symcc/package.nix:61]: Thanks. Fixed.

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**